File: blobs/howto/dotnet/dotnet-v12/CopyBlob.cs
using Azure;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using Azure.Storage.Blobs.Specialized;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace dotnet_v12
{
    class CopyBlob
    {
        //-------------------------------------------------
        // Copy a blob
        //-------------------------------------------------
        // <Snippet_CopyBlob>
        private static async Task CopyBlobAsync(BlobContainerClient container)
        {
            try
            {
                // Get the name of the first blob in the container to use as the source.
                string blobName = container.GetBlobs().FirstOrDefault().Name;

                // Create a BlobClient representing the source blob to copy.
                BlobClient sourceBlob = container.GetBlobClient(blobName);

                // Ensure that the source blob exists.
                if (await sourceBlob.ExistsAsync())
                {
                    // Lease the source blob for the copy operation 
                    // to prevent another client from modifying it.
                    BlobLeaseClient lease = sourceBlob.GetBlobLeaseClient();

                    // Specifying -1 for the lease interval creates an infinite lease.
                    await lease.AcquireAsync(TimeSpan.FromSeconds(-1));

                    // Get the source blob's properties and display the lease state.
                    BlobProperties sourceProperties = await sourceBlob.GetPropertiesAsync();
                    Console.WriteLine($"Lease state: {sourceProperties.LeaseState}");

                    // Get a BlobClient representing the destination blob with a unique name.
                    BlobClient destBlob = 
                        container.GetBlobClient(Guid.NewGuid() + "-" + sourceBlob.Name);

                    // Start the copy operation.
                    await destBlob.StartCopyFromUriAsync(sourceBlob.Uri);

                    // Get the destination blob's properties and display the copy status.
                    BlobProperties destProperties = await destBlob.GetPropertiesAsync();

                    Console.WriteLine($"Copy status: {destProperties.CopyStatus}");
                    Console.WriteLine($"Copy progress: {destProperties.CopyProgress}");
                    Console.WriteLine($"Completion time: {destProperties.CopyCompletedOn}");
                    Console.WriteLine($"Total bytes: {destProperties.ContentLength}");

                    // Update the source blob's properties.
                    sourceProperties = await sourceBlob.GetPropertiesAsync();

                    if (sourceProperties.LeaseState == LeaseState.Leased)
                    {
                        // Break the lease on the source blob.
                        await lease.BreakAsync();

                        // Update the source blob's properties to check the lease state.
                        sourceProperties = await sourceBlob.GetPropertiesAsync();
                        Console.WriteLine($"Lease state: {sourceProperties.LeaseState}");
                    }
                }
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine(ex.Message);
                Console.ReadLine();
                throw;
            }
        }
        // </Snippet_CopyBlob>

        //-------------------------------------------------
        // Stop a blob copy operation
        //-------------------------------------------------
        private static async Task StopBlobCopyAsync(BlobContainerClient container)
        {
            try
            {
                // Get the name of the first blob in the container to use as the source.
                string blobName = container.GetBlobs().FirstOrDefault().Name;

                // Create a BlobClient representing the source blob to copy.
                BlobClient sourceBlob = container.GetBlobClient(blobName);

                // Ensure that the source blob exists.
                if (await sourceBlob.ExistsAsync())
                {
                    // Get a BlobClient representing the destination blob with a unique name.
                    BlobClient destBlob = 
                        container.GetBlobClient(Guid.NewGuid() + "-" + sourceBlob.Name);

                    // Start the copy operation.
                    destBlob.StartCopyFromUri(sourceBlob.Uri);

                    // <Snippet_StopBlobCopy>
                    // Get the destination blob's properties to check the copy status.
                    BlobProperties destProperties = destBlob.GetProperties();

                    // Check the copy status. If the status is pending, abort the copy operation.
                    if (destProperties.CopyStatus == CopyStatus.Pending)
                    {
                        await destBlob.AbortCopyFromUriAsync(destProperties.CopyId);
                        Console.WriteLine($"Copy operation {destProperties.CopyId} has been aborted.");
                    }
                    // </Snippet_StopBlobCopy>
                    else
                    { 
                        Console.WriteLine($"Copy status: {destProperties.CopyStatus}");
                        Console.WriteLine($"Copy progress: {destProperties.CopyProgress}");
                        Console.WriteLine($"Completion time: {destProperties.CopyCompletedOn}");
                        Console.WriteLine($"Total bytes: {destProperties.ContentLength}");
                    }
                }
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine(ex.Message);
                Console.ReadLine();
                throw;
            }
        }

        public async Task<bool> MenuAsync()
        {
            var connectionString = Constants.connectionString;
            BlobServiceClient blobServiceClient = new BlobServiceClient(connectionString);
            BlobContainerClient container = blobServiceClient.GetBlobContainerClient(Constants.containerName);

            Console.Clear();
            Console.WriteLine("Choose a blob copy scenario:");
            Console.WriteLine("1) Copy an existing blob");
            Console.WriteLine("2) Stop blob copy operation");
            Console.WriteLine("X) Exit to main menu");
            Console.Write("\r\nSelect an option: ");

            switch (Console.ReadLine())
            {
                case "1":
                    await CopyBlobAsync(container);
                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "2":
                    await StopBlobCopyAsync(container);
                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "x":
                case "X":
                    return false;

                default:
                    return true;
            }
        }
    }
}

File: blobs/howto/python/python-v12/copy_blob.py
#----------------------------------------------------------------------------------
# Microsoft Developer & Platform Evangelism
#
# Copyright (c) Microsoft Corporation. All rights reserved.
#
# THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
# EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
# OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
#----------------------------------------------------------------------------------
# The example companies, organizations, products, domain names,
# e-mail addresses, logos, people, places, and events depicted
# herein are fictitious.  No association with any real company,
# organization, product, domain name, email address, logo, person,
# places, or events is intended or should be inferred.
#----------------------------------------------------------------------------------

# <Snippet_Imports>
import os, uuid

from azure.core.exceptions import (
    ResourceNotFoundError,
    ServiceRequestError
)

from azure.storage.blob import (
    BlobClient,
    BlobLeaseClient
)
# </Snippet_Imports>

from constants import Constants

class CopyBlob:
    def __init__(self):
        super().__init__()
        self.constants = Constants()

    # <Snippet_BlobCopy>
    def blob_copy(self, container_name, blob_name):

        # Create a BlobClient from a connection string
        # retrieved from an environment variable named
        # AZURE_STORAGE_CONNECTION_STRING.
        source_blob = BlobClient.from_connection_string(
            os.getenv("AZURE_STORAGE_CONNECTION_STRING"), 
            container_name, blob_name
            )

        try:
            # Lease the source blob for the copy operation
            # to prevent another client from modifying it.
            lease = BlobLeaseClient(source_blob)
            lease.acquire()

            # Get the source blob's properties and display the lease state.
            source_props = source_blob.get_blob_properties()
            print("Lease state: " + source_props.lease.state)

            # Create a BlobClient representing the
            # destination blob with a unique name.
            dest_blob = BlobClient.from_connection_string(
                os.getenv("AZURE_STORAGE_CONNECTION_STRING"),
                container_name, str(uuid.uuid4()) + "-" + blob_name
                )

            # Start the copy operation.
            dest_blob.start_copy_from_url(source_blob.url)

            # Get the destination blob's properties to check the copy status.
            properties = dest_blob.get_blob_properties()
            copy_props = properties.copy

            # Display the copy status.
            print("Copy status: " + copy_props["status"])
            print("Copy progress: " + copy_props["progress"])
            print("Completion time: " + str(copy_props["completion_time"]))
            print("Total bytes: " + str(properties.size))

            if (source_props.lease.state == "leased"):
                # Break the lease on the source blob.
                lease.break_lease()

                # Update the destination blob's properties to check the lease state.
                source_props = source_blob.get_blob_properties()
                print("Lease state: " + source_props.lease.state)

        except ResourceNotFoundError as ex:
            print("ResourceNotFoundError: ", ex.message)

        except ServiceRequestError as ex:
            print("ServiceRequestError: ", ex.message)
    # </Snippet_BlobCopy>

    def stop_blob_copy(self, container_name, blob_name):

        source_blob = BlobClient.from_connection_string(
            os.getenv("AZURE_STORAGE_CONNECTION_STRING"), 
            container_name, blob_name
            )

        try:
            dest_blob = BlobClient.from_connection_string(
                os.getenv("AZURE_STORAGE_CONNECTION_STRING"),
                container_name, str(uuid.uuid4()) + "-" + blob_name
                )

            # Start the copy operation.
            dest_blob.start_copy_from_url(source_blob.url)

            # <Snippet_StopBlobCopy>
            # Get the destination blob's properties to check the copy status.
            properties = dest_blob.get_blob_properties()
            copy_props = properties.copy

            # Check the copy status. If the status is pending, abort the copy operation.
            if (copy_props["status"] == "pending"):
                dest_blob.abort_copy(copy_props["id"])
                print("Copy operation " + copy_props["id"] + " has been aborted.")
            # </Snippet_StopBlobCopy>

            # Display the copy status.
            print("Copy status: " + copy_props["status"])
            print("Copy progress: " + copy_props["progress"])
            print("Completion time: " + str(copy_props["completion_time"]))
            print("Total bytes: " + str(properties.size))

        except ResourceNotFoundError as ex:
            print("ResourceNotFoundError: ", ex.message)

        except ServiceRequestError as ex:
            print("ServiceRequestError: ", ex.message)

    def menu(self):
        os.system("cls")
        print("Choose a blob copy scenario:")
        print("1) Copy an existing blob")
        print("2) Stop blob copy operation")
        print("X) Exit to main menu")
        option = input("\r\nSelect an option: ")

        if option == "1":
            self.blob_copy(self.constants.container_name, self.constants.blob_name)
            input("Press Enter to continue ")
            return True
        elif option == "2":
            self.stop_blob_copy(self.constants.container_name, self.constants.blob_name)
            input("Press Enter to continue ")
            return True
        elif option == "x" or option == "X":
            return False
        else:
            print("Unknown option: " + str(option))
            input("Press Enter to continue ")
            return True
